core/src/amp_subgraph/metrics.rs

```diff
@@ -25,12 +34,227 @@ impl Metrics {
+            indexing_duration: IndexingDuration::new(&metrics_registry, const_labels.clone()),
+            blocks_processed: BlocksProcessed::new(&metrics_registry, const_labels.clone()),
+            stopwatch,
+        }
+    }
+}
+
+/// Reports the current indexing status of a deployment.
+pub(super) struct DeploymentStatus(IntGauge);
+
+impl DeploymentStatus {
+    const STATUS_STARTING: i64 = 1;
+    const STATUS_RUNNING: i64 = 2;
+    const STATUS_STOPPED: i64 = 3;
+    const STATUS_FAILED: i64 = 4;
+
+    fn new(
+        metrics_registry: &MetricsRegistry,
+        const_labels: impl IntoIterator<Item = (impl ToString, impl ToString)>,
+    ) -> Self {
+        let int_gauge = metrics_registry
+            .new_int_gauge(
+                "amp_deployment_status",
+                indoc!(
+                    "
+                    Indicates the current indexing status of a deployment.
+                    Possible values:
+                    1 - graph-node is preparing to start indexing;
+                    2 - deployment is being indexed;
+                    3 - indexing is stopped by request;
+                    4 - indexing failed;
+                    "
+                ),
+                const_labels,
+            )
+            .expect("failed to register `amp_deployment_status` gauge");
+
+        Self(int_gauge)
+    }
+
+    /// Records that the graph-node is preparing to start indexing.
+    pub fn starting(&self) {
+        self.0.set(Self::STATUS_STARTING);
+    }
+
+    /// Records that the deployment is being indexed.
+    pub fn running(&self) {
+        self.0.set(Self::STATUS_RUNNING);
+    }
+
+    /// Records that the indexing stopped by request.
+    pub fn stopped(&self) {
+        self.0.set(Self::STATUS_STOPPED);
+    }
+
+    /// Records that the indexing failed.
+    pub fn failed(&self) {
+        self.0.set(Self::STATUS_FAILED);
+    }
+}
+
+/// Tracks the most recent block number processed by a deployment.
+pub(super) struct DeploymentHead(IntGauge);
+
+impl DeploymentHead {
+    fn new(
+        metrics_registry: &MetricsRegistry,
+        const_labels: impl IntoIterator<Item = (impl ToString, impl ToString)>,
+    ) -> Self {
+        let int_gauge = metrics_registry
+            .new_int_gauge(
+                "amp_deployment_head",
+                "Tracks the most recent block number processed by a deployment",
+                const_labels,
+            )
+            .expect("failed to register `amp_deployment_head` gauge");
+
+        Self(int_gauge)
+    }
+
+    /// Updates the most recent block number processed by this deployment.
+    pub(super) fn update(&self, new_most_recent_block_number: BlockNumber) {
+        self.0.set(
+            i64::try_from(new_most_recent_block_number)
+                .expect("new most recent block number does not fit into `i64`"),
+        );
+    }
+}
+
+/// Tracks the target block number of a deployment.
+pub(super) struct DeploymentTarget(IntGauge);
```

Comment:
I don't understand what 'target' means here - is this basically the latest block that Amp has available?